Segfaults and test failures under osx-64

RIAK 0.2.2 often throws segfaults on an Intel Mac and fails tests. Does not appear to happen with every input file, although it is not restricted to debug mode as per previous issue, although that certainly does appear problematic.

## Preamble
```
conda create -n riaktest python=3 pytest pyfastaq art badread
conda activate riaktest
cd src
make
make test
```

## Conda env spec
```
% conda env export
name: riaktest
channels:
  - conda-forge
  - bioconda
  - defaults
dependencies:
  - art=2016.06.05=h9ca5826_8
  - attrs=21.4.0=pyhd8ed1ab_0
  - badread=0.2.0=py_0
  - brotli=1.0.9=h5eb16cf_7
  - brotli-bin=1.0.9=h5eb16cf_7
  - bzip2=1.0.8=h0d85af4_4
  - ca-certificates=2022.3.29=hecd8cb5_0
  - certifi=2021.10.8=py310h2ec42d9_2
  - cycler=0.11.0=pyhd8ed1ab_0
  - edlib=1.3.9=py310h2fea185_1
  - fonttools=4.32.0=py310h1961e1f_0
  - freetype=2.11.0=hd8bbffd_0
  - giflib=5.2.1=hbcb3906_2
  - gsl=2.7.1=hdbe807d_0
  - iniconfig=1.1.1=pyh9f0ad1d_0
  - jbig=2.1=h0d85af4_2003
  - jpeg=9e=h5eb16cf_1
  - kiwisolver=1.4.2=py310h6be76da_1
  - lcms2=2.12=h577c468_0
  - lerc=3.0=he49afe7_0
  - libblas=3.9.0=14_osx64_openblas
  - libbrotlicommon=1.0.9=h5eb16cf_7
  - libbrotlidec=1.0.9=h5eb16cf_7
  - libbrotlienc=1.0.9=h5eb16cf_7
  - libcblas=3.9.0=14_osx64_openblas
  - libcxx=13.0.1=hc203e6f_0
  - libdeflate=1.10=h0d85af4_0
  - libffi=3.4.2=h0d85af4_5
  - libgfortran=5.0.0=9_3_0_h6c81a4c_23
  - libgfortran5=9.3.0=h6c81a4c_23
  - liblapack=3.9.0=14_osx64_openblas
  - libopenblas=0.3.20=openmp_hb3cd9ec_0
  - libpng=1.6.37=h7cec526_2
  - libtiff=4.3.0=h17f2ce3_3
  - libwebp=1.2.2=h28dabe5_0
  - libwebp-base=1.2.2=h0d85af4_1
  - libxcb=1.13=h0d85af4_1004
  - libzlib=1.2.11=h6c3fc93_1014
  - llvm-openmp=13.0.1=hcb1a161_1
  - lz4-c=1.9.3=he49afe7_1
  - matplotlib-base=3.5.1=py310hfc59cd6_0
  - munkres=1.1.4=pyh9f0ad1d_0
  - ncurses=6.3=h96cf925_1
  - numpy=1.22.3=py310hb608527_2
  - openjpeg=2.4.0=h6e7aa92_1
  - openssl=3.0.2=h6c3fc93_1
  - packaging=21.3=pyhd8ed1ab_0
  - pillow=9.1.0=py310h382dc3b_2
  - pip=22.0.4=pyhd8ed1ab_0
  - pluggy=1.0.0=py310h2ec42d9_3
  - pthread-stubs=0.4=hc929b4f_1001
  - py=1.11.0=pyh6c4a22f_0
  - pyfastaq=3.17.0=py_2
  - pyparsing=3.0.8=pyhd8ed1ab_0
  - pytest=7.1.1=py310h2ec42d9_1
  - python=3.10.4=h1cc4136_0_cpython
  - python-dateutil=2.8.2=pyhd8ed1ab_0
  - python_abi=3.10=2_cp310
  - readline=8.1.2=hca72f7f_1
  - scipy=1.8.0=py310h47774c9_1
  - setuptools=62.1.0=py310h2ec42d9_0
  - six=1.16.0=pyh6c4a22f_0
  - sqlite=3.38.2=hb516253_0
  - tk=8.6.12=h5dbffcc_0
  - tomli=2.0.1=pyhd8ed1ab_0
  - tzdata=2022a=h191b570_0
  - unicodedata2=14.0.0=py310h1961e1f_1
  - wheel=0.37.1=pyhd8ed1ab_0
  - xorg-libxau=1.0.9=h35c211d_0
  - xorg-libxdmcp=1.1.3=h35c211d_0
  - xz=5.2.5=haf1e3a3_1
  - zlib=1.2.11=h6c3fc93_1014
  - zstd=1.5.2=h582d3a0_0
prefix: /Users/bede/miniconda3/envs/riaktest
```

## Test output
```
(riaktest) bede@yoshina src % make test
cd ../tests && pytest
======================================================================================================== test session starts =========================================================================================================
platform darwin -- Python 3.10.4, pytest-7.1.1, pluggy-1.0.0
rootdir: /Users/bede/Downloads/read-it-and-keep-0.2.2/tests
collected 7 items                                                                                                                                                                                                                    

test_all.py FF..FFF                                                                                                                                                                                                            [100%]

============================================================================================================== FAILURES ==============================================================================================================
______________________________________________________________________________________________________ test_shredded_ref_genome ______________________________________________________________________________________________________

    def test_shredded_ref_genome():
        outprefix = "out.shredded_ref_genome"
        subprocess.check_output(f"rm -rf {outprefix}*", shell=True)
        shred_length = 75
        shredded_ref = f"{outprefix}.k-{shred_length}.fa"
        shred_ref_genome(shredded_ref, shred_length)
        riak_out = f"{outprefix}.riak"
>       riak_counts = run_read_it_and_keep([shredded_ref], riak_out, "illumina")

test_all.py:158: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_all.py:61: in run_read_it_and_keep
    subprocess.check_output(command, shell=True)
../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:420: in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True
popenargs = ('../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.shredded_ref_genome.k-75.fa -o out.shredded_ref_genome.riak > out.shredded_ref_genome.riak.out',)
kwargs = {'shell': True, 'stdout': -1}, process = <Popen: returncode: 139 args: '../src/readItAndKeep --debug  --tech illumina...>, stdout = b'', stderr = None, retcode = 139

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.shredded_ref_genome.k-75.fa -o out.shredded_ref_genome.riak > out.shredded_ref_genome.riak.out' returned non-zero exit status 139.

../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:524: CalledProcessError
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
/bin/sh: line 1: 81691 Segmentation fault: 11  ../src/readItAndKeep --debug --tech illumina --ref_fasta MN908947.3.fa --reads1 out.shredded_ref_genome.k-75.fa -o out.shredded_ref_genome.riak > out.shredded_ref_genome.riak.out
______________________________________________________________________________________________________ test_illumina_hiseq_sim _______________________________________________________________________________________________________

    def test_illumina_hiseq_sim():
        outprefix = "out.illumina_hiseq_sim"
        subprocess.check_output(f"rm -rf {outprefix}*", shell=True)
        illumina_read_prefix = f"{outprefix}.reads"
>       simulate_illumina(illumina_read_prefix)

test_all.py:180: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_all.py:133: in simulate_illumina
    subprocess.check_output(command, shell=True)
../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:420: in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True, popenargs = ('art_illumina --in MN908947.3.fa --out out.illumina_hiseq_sim.reads --noALN --seqSys HS25 --len 150 --fcov 10 --mflen 250 --sdev 10 --rndSeed 42',)
kwargs = {'shell': True, 'stdout': -1}, process = <Popen: returncode: -6 args: 'art_illumina --in MN908947.3.fa --out out.illu...>, stdout = b'', stderr = None, retcode = -6

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command 'art_illumina --in MN908947.3.fa --out out.illumina_hiseq_sim.reads --noALN --seqSys HS25 --len 150 --fcov 10 --mflen 250 --sdev 10 --rndSeed 42' died with <Signals.SIGABRT: 6>.

../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:524: CalledProcessError
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
dyld[81693]: Library not loaded: @rpath/libgsl.25.dylib
  Referenced from: /Users/bede/miniconda3/envs/riaktest/bin/art_illumina
  Reason: tried: '/Users/bede/miniconda3/envs/riaktest/bin/../lib/libgsl.25.dylib' (no such file), '/Users/bede/miniconda3/envs/riaktest/bin/../lib/libgsl.25.dylib' (no such file), '/usr/local/lib/libgsl.25.dylib' (no such file), '/usr/lib/libgsl.25.dylib' (no such file)
______________________________________________________________________________________________________ test_delta_382_deletion _______________________________________________________________________________________________________

    def test_delta_382_deletion():
        outprefix = "out.delta_382_deletion"
        subprocess.check_output(f"rm -rf {outprefix}*", shell=True)
        sim_reads_fasta = f"{outprefix}.for_reads.fa"
        make_ref_genome_with_deletion(sim_reads_fasta, 27848, 28229, 26848, 29229)
    
        nano_reads = f"{outprefix}.reads.fq"
        simulate_nanopore(nano_reads, read_length_stdev="1000,50", ref=sim_reads_fasta)
        number_of_input_reads = 32
        assert pyfastaq.tasks.count_sequences(nano_reads) == number_of_input_reads
        riak_out = f"{outprefix}.riak"
        riak_counts = run_read_it_and_keep([nano_reads], riak_out, "ont")
        reads_out = f"{riak_out}.reads.fastq.gz"
        # Some short reads/low % identity reads are not kept. This is ok. Their descriptions are:
        # MN908947.3,+strand,1807-2703 length=219 error-free_length=220 read_identity=86.83%
        # MN908947.3,-strand,1946-3027 length=78 error-free_length=80 read_identity=86.46%
        # MN908947.3,+strand,1721-2636 length=286 error-free_length=314 read_identity=73.39%
        # MN908947.3,+strand,1985-2995 length=36 error-free_length=38 read_identity=90.57%
        # junk_seq length=997 error-free_length=1022 read_identity=83.73%
        # MN908947.3,+strand,5-982 length=922 error-free_length=980 read_identity=76.17%
    
        number_of_output_reads = 26
        assert os.path.exists(reads_out)
        assert riak_debug_check_badreads_failed_reads_ok(f"{riak_out}.reads.debug")
        assert pyfastaq.tasks.count_sequences(reads_out) == number_of_output_reads
        assert riak_counts["Input reads file 1"] == number_of_input_reads
        assert riak_counts["Input reads file 2"] == 0
        assert riak_counts["Kept reads 1"] == number_of_output_reads
        assert riak_counts["Kept reads 2"] == 0
        subprocess.check_output(f"rm -rf {nano_reads} {reads_out} {riak_out}*", shell=True)
    
        illumina_read_prefix = f"{outprefix}.reads"
>       simulate_illumina(illumina_read_prefix, ref=sim_reads_fasta)

test_all.py:285: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_all.py:133: in simulate_illumina
    subprocess.check_output(command, shell=True)
../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:420: in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True
popenargs = ('art_illumina --in out.delta_382_deletion.for_reads.fa --out out.delta_382_deletion.reads --noALN --seqSys HS25 --len 150 --fcov 10 --mflen 250 --sdev 10 --rndSeed 42',), kwargs = {'shell': True, 'stdout': -1}
process = <Popen: returncode: -6 args: 'art_illumina --in out.delta_382_deletion.for_r...>, stdout = b'', stderr = None, retcode = -6

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command 'art_illumina --in out.delta_382_deletion.for_reads.fa --out out.delta_382_deletion.reads --noALN --seqSys HS25 --len 150 --fcov 10 --mflen 250 --sdev 10 --rndSeed 42' died with <Signals.SIGABRT: 6>.

../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:524: CalledProcessError
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------

Badread v0.2.0
long read simulation

Loading reference from out.delta_382_deletion.for_reads.fa
  1 contig:
    MN908947.3: 2,000 bp, linear, 1.00x depth

Generating fragment lengths from a gamma distribution:
  mean  =   1000 bp      parameters:
  stdev =     50 bp        k (shape)     = 4.0000e+02
  N50   =   1002 bp        theta (scale) = 2.5000e+00
  │            ▌                                     
  │            ▌                                     
f │            ▌                                     
r │            ▌                                     
a │            ▌                                     
g │            ▌                                     
s │           ▌▌▌                                    
  │           ▌▌▌                                    
  ├─────────┬─────────┬─────────┬─────────┬─────────┐
  │            ▌                                     
  │            ▌                                     
b │            ▌                                     
a │            ▌                                     
s │            ▌                                     
e │            ▌▖                                    
s │           ▌▌▌                                    
  │           ▌▌▌                                    
  ├─────────┬─────────┬─────────┬─────────┬─────────┐
  0         800       1600      2400      3200      4000

Generating read identities from a beta distribution:
  mean  = 87.5%      shape parameters:
  max   = 97.5%        alpha = 3.0513e+01
  stdev =   5%         beta  = 3.4872e+00
  │                                      ▖▌▌▖        
  │                                    ▖▌▌▌▌▌▖       
  │                                   ▖▌▌▌▌▌▌▌       
  │                                  ▌▌▌▌▌▌▌▌▌▌      
  │                                ▖▌▌▌▌▌▌▌▌▌▌▌▖     
  │                              ▖▌▌▌▌▌▌▌▌▌▌▌▌▌▌     
  │                            ▖▌▌▌▌▌▌▌▌▌▌▌▌▌▌▌▌▖    
  │                      ▖▖▖▖▌▌▌▌▌▌▌▌▌▌▌▌▌▌▌▌▌▌▌▌▖   
  ├─────────┬─────────┬─────────┬─────────┬─────────┐
  50        60        70        80        90        100

Loading error model from /Users/bede/miniconda3/envs/riaktest/lib/python3.10/site-packages/badread/error_models/nanopore2020.gz
  done: loaded error distributions for 16384 7-mers

Loading qscore model from /Users/bede/miniconda3/envs/riaktest/lib/python3.10/site-packages/badread/qscore_models/nanopore2020.gz
  done: loaded qscore distributions for 10000 alignments

Read glitches:
  rate (mean distance between glitches) = 10000
  size (mean length of random sequence) =    25
  skip (mean sequence lost per glitch)  =    25

Start adapter:
  seq: AATGTACTTCGTTCAGTTACGTATTGCT
  rate:   90.0%
  amount: 60.0%

End adapter:
  seq: GCAATACGTAACTGAACGAAGT
  rate:   50.0%
  amount: 20.0%

Other problems:
  chimera join rate: 1%
  junk read rate:    1%
  random read rate:  1%

Target read set size: 20,000 bp

Simulating: 32 reads  20,325 bp  100.0%

dyld[81725]: Library not loaded: @rpath/libgsl.25.dylib
  Referenced from: /Users/bede/miniconda3/envs/riaktest/bin/art_illumina
  Reason: tried: '/Users/bede/miniconda3/envs/riaktest/bin/../lib/libgsl.25.dylib' (no such file), '/Users/bede/miniconda3/envs/riaktest/bin/../lib/libgsl.25.dylib' (no such file), '/usr/local/lib/libgsl.25.dylib' (no such file), '/usr/lib/libgsl.25.dylib' (no such file)
___________________________________________________________________________________________________ test_enumerate_names_unpaired ____________________________________________________________________________________________________

    def test_enumerate_names_unpaired():
        outprefix = "out.enumerate_names_unpaired"
        subprocess.check_output(f"rm -rf {outprefix}*", shell=True)
        read1 = COVID_REF_SEQ[0:100]
        read2 = COVID_REF_SEQ[200:300]
        test_reads_fa = f"{outprefix}.reads.fq"
        with open(test_reads_fa, "w") as f:
            print(">read1", read1, sep="\n", file=f)
            print(">read2", read2, sep="\n", file=f)
        riak_out = f"{outprefix}.riak"
    
>       run_read_it_and_keep([test_reads_fa], outprefix, "illumina", enumerate_names=False)

test_all.py:318: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_all.py:61: in run_read_it_and_keep
    subprocess.check_output(command, shell=True)
../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:420: in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True
popenargs = ('../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.enumerate_names_unpaired.reads.fq -o out.enumerate_names_unpaired > out.enumerate_names_unpaired.out',)
kwargs = {'shell': True, 'stdout': -1}, process = <Popen: returncode: 139 args: '../src/readItAndKeep --debug  --tech illumina...>, stdout = b'', stderr = None, retcode = 139

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.enumerate_names_unpaired.reads.fq -o out.enumerate_names_unpaired > out.enumerate_names_unpaired.out' returned non-zero exit status 139.

../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:524: CalledProcessError
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
/bin/sh: line 1: 81728 Segmentation fault: 11  ../src/readItAndKeep --debug --tech illumina --ref_fasta MN908947.3.fa --reads1 out.enumerate_names_unpaired.reads.fq -o out.enumerate_names_unpaired > out.enumerate_names_unpaired.out
____________________________________________________________________________________________________ test_enumerate_names_paired _____________________________________________________________________________________________________

    def test_enumerate_names_paired():
        outprefix = "out.enumerate_names_paired"
        subprocess.check_output(f"rm -rf {outprefix}*", shell=True)
        read1 = COVID_REF_SEQ[0:100]
        read2 = COVID_REF_SEQ[200:300]
        read3 = COVID_REF_SEQ[300:400]
        read4 = COVID_REF_SEQ[400:500]
        test_reads_fa1 = f"{outprefix}.reads.1.fq"
        test_reads_fa2 = f"{outprefix}.reads.2.fq"
        with open(test_reads_fa1, "w") as f:
            print(">read1", read1, sep="\n", file=f)
            print(">read2", read2, sep="\n", file=f)
        with open(test_reads_fa2, "w") as f:
            print(">read1", read3, sep="\n", file=f)
            print(">read2", read4, sep="\n", file=f)
        riak_out = f"{outprefix}.riak"
    
        fq_list = [test_reads_fa1, test_reads_fa2]
>       run_read_it_and_keep(fq_list, outprefix, "illumina", enumerate_names=False)

test_all.py:345: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_all.py:61: in run_read_it_and_keep
    subprocess.check_output(command, shell=True)
../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:420: in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True
popenargs = ('../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.enumerate_names_paired.reads.1.fq --reads2 out.enumerate_names_paired.reads.2.fq -o out.enumerate_names_paired > out.enumerate_names_paired.out',)
kwargs = {'shell': True, 'stdout': -1}, process = <Popen: returncode: 139 args: '../src/readItAndKeep --debug  --tech illumina...>, stdout = b'', stderr = None, retcode = 139

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.enumerate_names_paired.reads.1.fq --reads2 out.enumerate_names_paired.reads.2.fq -o out.enumerate_names_paired > out.enumerate_names_paired.out' returned non-zero exit status 139.

../../../miniconda3/envs/riaktest/lib/python3.10/subprocess.py:524: CalledProcessError
-------------------------------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------------------------------
/bin/sh: line 1: 81731 Segmentation fault: 11  ../src/readItAndKeep --debug --tech illumina --ref_fasta MN908947.3.fa --reads1 out.enumerate_names_paired.reads.1.fq --reads2 out.enumerate_names_paired.reads.2.fq -o out.enumerate_names_paired > out.enumerate_names_paired.out
====================================================================================================== short test summary info =======================================================================================================
FAILED test_all.py::test_shredded_ref_genome - subprocess.CalledProcessError: Command '../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.shredded_ref_genome.k-75.fa -o out.shredded_ref_genome....
FAILED test_all.py::test_illumina_hiseq_sim - subprocess.CalledProcessError: Command 'art_illumina --in MN908947.3.fa --out out.illumina_hiseq_sim.reads --noALN --seqSys HS25 --len 150 --fcov 10 --mflen 250 --sdev 10 --rndSeed ...
FAILED test_all.py::test_delta_382_deletion - subprocess.CalledProcessError: Command 'art_illumina --in out.delta_382_deletion.for_reads.fa --out out.delta_382_deletion.reads --noALN --seqSys HS25 --len 150 --fcov 10 --mflen 25...
FAILED test_all.py::test_enumerate_names_unpaired - subprocess.CalledProcessError: Command '../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.enumerate_names_unpaired.reads.fq -o out.enumerate...
FAILED test_all.py::test_enumerate_names_paired - subprocess.CalledProcessError: Command '../src/readItAndKeep --debug  --tech illumina --ref_fasta MN908947.3.fa --reads1 out.enumerate_names_paired.reads.1.fq --reads2 out.enume...
==================================================================================================== 5 failed, 2 passed in 26.53s ====================================================================================================
make: *** [test] Error 1
(riaktest) bede@yoshina src %                                                      

```

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -125,6 +125,14 @@ jobs:
     - name: Install dependencies
       run: |
         pip3 install pytest pyfastaq
+        wget https://www.niehs.nih.gov/research/resources/assets/docs/artbinmountrainier2016.06.05linux64.tgz
+        tar xf artbinmountrainier2016.06.05linux64.tgz
+        rm artbinmountrainier2016.06.05linux64.tgz
+        wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
+        tar xf v0.2.0.tar.gz
+        rm v0.2.0.tar.gz
+        cd Badread-0.2.0
+        pip3 install .
 
     - name: Check out code for the build
       uses: actions/checkout@v2
@@ -133,3 +141,6 @@ jobs:
       run: |
           cd src
           make
+          export PATH=../art_bin_MountRainier:$PATH
+          make test
+

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -137,6 +137,13 @@ jobs:
         wget -q https://www.niehs.nih.gov/research/resources/assets/docs/artsrcmountrainier2016.06.05macos.tgz
         tar xf artsrcmountrainier2016.06.05macos.tgz
         rm artsrcmountrainier2016.06.05macos.tgz
+        cd art_src_MountRainier_MacOS
+        export CFLAGS="$CFLAGS -I/usr/local/include" CPPFLAGS="$CPPFLAGS -I/usr/local/include" LDFLAGS="$LDFLAGS -L/usr/local/lib"
+        echo "ART: configure"
+        ./configure
+        echo "ART: make"
+        make
+        cd ..
         echo "pwd"
         pwd
         echo "ls -lh"

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -6,7 +6,6 @@ on:
       - 'v*.*.*'
     branches:
       - main
-      - github_actions_mac
   pull_request:
     branches:
       - main

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -155,7 +155,7 @@ jobs:
         export PATH=$PWD/art_src_MountRainier_MacOS:$PATH
         echo "PATH:"
         echo $PATH
-        echo "art_illimina --help"
+        echo "art_illumina --help"
         art_illumina --help | cat
 
         wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
@@ -171,6 +171,12 @@ jobs:
       run: |
           echo "pwd:"
           pwd
+          echo "ls -lh"
+          ls -lh
+          echo "ls -lh art_src_MountRainier_MacOS"
+          ls -lh art_src_MountRainier_MacOS
+          echo "find . -name art_illumina"
+          find . -name art_illumina
           echo "PATH:"
           echo $PATH
           export PATH=$PWD/art_src_MountRainier_MacOS:$PATH

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -165,7 +165,7 @@ jobs:
           pwd
           echo "PATH:"
           echo $PATH
-          export PATH=$PWD/art_bin_MountRainier_MacOS:$PATH
+          export PATH=$PWD/art_src_MountRainier_MacOS:$PATH
           echo "PATH:"
           echo $PATH
           cd src

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -150,8 +150,6 @@ jobs:
         ls -lh
         echo "ls -lh art_src_MountRainier_MacOS"
         ls -lh art_src_MountRainier_MacOS
-        echo "./art_src_MountRainier_MacOS/art_illumina"
-        ./art_src_MountRainier_MacOS/art_illumina
         wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
         tar xf v0.2.0.tar.gz
         rm v0.2.0.tar.gz

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -142,24 +142,10 @@ jobs:
         rm artsrcmountrainier2016.06.05macos.tgz
         cd art_src_MountRainier_MacOS
         export CFLAGS="$CFLAGS -I/usr/local/include" CPPFLAGS="$CPPFLAGS -I/usr/local/include" LDFLAGS="$LDFLAGS -L/usr/local/lib"
-        echo "ART: configure"
         ./configure
-        echo "ART: make"
         make
         cd ..
-        echo "pwd"
         pwd
-        echo "ls -lh"
-        ls -lh
-        echo "ls -lh art_src_MountRainier_MacOS"
-        ls -lh art_src_MountRainier_MacOS
-        echo "PATH:"
-        echo $PATH
-        export PATH=$PWD/art_src_MountRainier_MacOS:$PATH
-        echo "PATH:"
-        echo $PATH
-        echo "art_illumina --help"
-        art_illumina --help | cat
 
         wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
         tar xf v0.2.0.tar.gz
@@ -169,21 +155,8 @@ jobs:
 
     - name: Build and run tests
       run: |
-          echo "pwd:"
-          pwd
-          echo "ls -lh"
-          ls -lh
-          echo "ls -lh art_src_MountRainier_MacOS"
-          ls -lh art_src_MountRainier_MacOS
-          echo "find . -name art_illumina"
-          find . -name art_illumina
-          echo "PATH:"
-          echo $PATH
           export PATH=$PWD/art_src_MountRainier_MacOS:$PATH
-          echo "PATH:"
-          echo $PATH
           cd src
-          art_illumina --help | cat
           make
           make test
 

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -125,15 +125,17 @@ jobs:
     - name: Install dependencies
       run: |
         pip3 install pytest pyfastaq
-        wget -q https://www.niehs.nih.gov/research/resources/assets/docs/artbinmountrainier2016.06.05linux64.tgz
-        tar xf artbinmountrainier2016.06.05linux64.tgz
-        rm artbinmountrainier2016.06.05linux64.tgz
+        wget -q https://www.niehs.nih.gov/research/resources/assets/docs/artsrcmountrainier2016.06.05macos.tgz
+        tar xf artsrcmountrainier2016.06.05macos.tgz
+        rm artsrcmountrainier2016.06.05macos.tgz
         echo "pwd"
         pwd
         echo "ls -lh"
         ls -lh
-        echo "ls -lh art_bin_MountRainier"
-        ls -lh art_bin_MountRainier
+        echo "ls -lh art_src_MountRainier_MacOS"
+        ls -lh art_src_MountRainier_MacOS
+        echo "./art_src_MountRainier_MacOS/art_illumina"
+        ./art_src_MountRainier_MacOS/art_illumina
         wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
         tar xf v0.2.0.tar.gz
         rm v0.2.0.tar.gz
@@ -149,10 +151,12 @@ jobs:
           pwd
           echo "PATH:"
           echo $PATH
-          export PATH=$PWD/art_bin_MountRainier:$PATH
+          export PATH=$PWD/art_bin_MountRainier_MacOS:$PATH
           echo "PATH:"
           echo $PATH
           cd src
+          echo "which art_illumina"
+          which art_illumina
           make
           make test
 

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -6,6 +6,7 @@ on:
       - 'v*.*.*'
     branches:
       - main
+      - github_actions_mac
   pull_request:
     branches:
       - main
@@ -16,7 +17,7 @@ env:
 
 jobs:
   build:
-    name: Build
+    name: Build linux/singularity
     runs-on: ubuntu-20.04
     steps:
 
@@ -114,3 +115,21 @@ jobs:
         push: true
         tags: ${{ steps.meta.outputs.tags }}
         labels: ${{ steps.meta.outputs.labels }}
+
+
+  mac_build:
+    name: Build mac os
+    runs-on: macos-11
+    steps:
+
+    - name: Install dependencies
+      run: |
+        pip3 install pytest pyfastaq
+
+    - name: Check out code for the build
+      uses: actions/checkout@v2
+
+    - name: Build and run tests
+      run: |
+          cd src
+          make

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -139,8 +139,8 @@ jobs:
 
     - name: Build and run tests
       run: |
+          export PATH=$PWD/art_bin_MountRainier:$PATH
           cd src
           make
-          export PATH=../art_bin_MountRainier:$PATH
           make test
 

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -122,6 +122,9 @@ jobs:
     runs-on: macos-11
     steps:
 
+    - name: Check out code for the build
+      uses: actions/checkout@v2
+
     - name: Install dependencies
       run: |
         pip3 install pytest pyfastaq
@@ -163,9 +166,6 @@ jobs:
         rm v0.2.0.tar.gz
         cd Badread-0.2.0
         pip3 install .
-
-    - name: Check out code for the build
-      uses: actions/checkout@v2
 
     - name: Build and run tests
       run: |

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -150,6 +150,16 @@ jobs:
         ls -lh
         echo "ls -lh art_src_MountRainier_MacOS"
         ls -lh art_src_MountRainier_MacOS
+        echo "./art_src_MountRainier_MacOS/art_illumina --help"
+        ./art_src_MountRainier_MacOS/art_illumina --help
+        echo "PATH:"
+        echo $PATH
+        export PATH=$PWD/art_src_MountRainier_MacOS:$PATH
+        echo "PATH:"
+        echo $PATH
+        echo "art_illimina --help"
+        art_illumina --help
+
         wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
         tar xf v0.2.0.tar.gz
         rm v0.2.0.tar.gz
@@ -169,8 +179,17 @@ jobs:
           echo "PATH:"
           echo $PATH
           cd src
-          echo "which art_illumina"
-          which art_illumina
+          art_illumina --help
           make
           make test
 
+
+
+
+
+          /Users/runner/work/read-it-and-keep/read-it-and-keep
+ls -lh
+total 0
+drwxr-xr-x   42 runner  staff   1.3K Apr 20 17:01 art_src_MountRainier_MacOS
+drwxr-xr-x  128 runner  staff   4.0K Apr 20 17:00 gsl-2.7.1
+ls -lh art_src_MountRainier_MacOS

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -132,7 +132,7 @@ jobs:
         pwd
         echo "ls -lh"
         ls -lh
-        "echo ls -lh art_bin_MountRainier"
+        echo "ls -lh art_bin_MountRainier"
         ls -lh art_bin_MountRainier
         wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
         tar xf v0.2.0.tar.gz
@@ -145,6 +145,8 @@ jobs:
 
     - name: Build and run tests
       run: |
+          echo "pwd:"
+          pwd
           echo "PATH:"
           echo $PATH
           export PATH=$PWD/art_bin_MountRainier:$PATH

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -125,9 +125,15 @@ jobs:
     - name: Install dependencies
       run: |
         pip3 install pytest pyfastaq
-        wget https://www.niehs.nih.gov/research/resources/assets/docs/artbinmountrainier2016.06.05linux64.tgz
+        wget -q https://www.niehs.nih.gov/research/resources/assets/docs/artbinmountrainier2016.06.05linux64.tgz
         tar xf artbinmountrainier2016.06.05linux64.tgz
         rm artbinmountrainier2016.06.05linux64.tgz
+        echo "pwd"
+        pwd
+        echo "ls -lh"
+        ls -lh
+        "echo ls -lh art_bin_MountRainier"
+        ls -lh art_bin_MountRainier
         wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
         tar xf v0.2.0.tar.gz
         rm v0.2.0.tar.gz
@@ -139,7 +145,11 @@ jobs:
 
     - name: Build and run tests
       run: |
+          echo "PATH:"
+          echo $PATH
           export PATH=$PWD/art_bin_MountRainier:$PATH
+          echo "PATH:"
+          echo $PATH
           cd src
           make
           make test

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -125,6 +125,15 @@ jobs:
     - name: Install dependencies
       run: |
         pip3 install pytest pyfastaq
+        wget -q https://ftp.gnu.org/gnu/gsl/gsl-2.7.1.tar.gz
+        tar xf gsl-2.7.1.tar.gz
+        rm gsl-2.7.1.tar.gz
+        cd gsl-2.7.1
+        ./configure
+        make
+        make install
+        cd ..
+
         wget -q https://www.niehs.nih.gov/research/resources/assets/docs/artsrcmountrainier2016.06.05macos.tgz
         tar xf artsrcmountrainier2016.06.05macos.tgz
         rm artsrcmountrainier2016.06.05macos.tgz

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -150,15 +150,13 @@ jobs:
         ls -lh
         echo "ls -lh art_src_MountRainier_MacOS"
         ls -lh art_src_MountRainier_MacOS
-        echo "./art_src_MountRainier_MacOS/art_illumina --help"
-        ./art_src_MountRainier_MacOS/art_illumina --help
         echo "PATH:"
         echo $PATH
         export PATH=$PWD/art_src_MountRainier_MacOS:$PATH
         echo "PATH:"
         echo $PATH
         echo "art_illimina --help"
-        art_illumina --help
+        art_illumina --help | cat
 
         wget -q https://github.com/rrwick/Badread/archive/refs/tags/v0.2.0.tar.gz
         tar xf v0.2.0.tar.gz
@@ -179,7 +177,7 @@ jobs:
           echo "PATH:"
           echo $PATH
           cd src
-          art_illumina --help
+          art_illumina --help | cat
           make
           make test
 

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -183,13 +183,3 @@ jobs:
           make
           make test
 
-
-
-
-
-          /Users/runner/work/read-it-and-keep/read-it-and-keep
-ls -lh
-total 0
-drwxr-xr-x   42 runner  staff   1.3K Apr 20 17:01 art_src_MountRainier_MacOS
-drwxr-xr-x  128 runner  staff   4.0K Apr 20 17:00 gsl-2.7.1
-ls -lh art_src_MountRainier_MacOS

--- a/src/readItAndKeep.cpp
+++ b/src/readItAndKeep.cpp
@@ -328,7 +328,7 @@ void QueryReads::write(unsigned long long int readNumber) {
     }
     else {
         filehandleOut_ << headerChar_ << readPtr->name.s;
-        if (readPtr->comment.l > 0 and not isBeingRenamed_) {
+        if (readPtr->comment.s != 0 and readPtr->comment.l > 0 and not isBeingRenamed_) {
             filehandleOut_ << ' ' << readPtr->comment.s;
         }
     }
@@ -348,7 +348,7 @@ void QueryReads::writeDebug() {
         return;
     }
     filehandleDebugOut_ << "REJECTED_READ\t" << readPtr->name.s
-                        << '\t' << readPtr->comment.s
+                        << '\t' << ((readPtr->comment.s != 0 and readPtr->comment.l > 0) ? readPtr->comment.s : "NA")
                         << '\t' << readPtr->seq.s
                         << '\t' << readPtr->qual.s << '\n';
 }
@@ -357,7 +357,7 @@ bool QueryReads::isGoodMapping() {
     int j;
     if (debug_) {
         filehandleDebugOut_ << "MAPPING_COUNT\t" << readPtr->name.s
-                            << '\t' << readPtr->comment.s
+                            << '\t' << ((readPtr->comment.s != 0 and readPtr->comment.l > 0) ? readPtr->comment.s : "NA")
                             << '\t' << n_reg_ << '\n';
     }
     for (j = 0; j < n_reg_; ++j) {
@@ -367,7 +367,7 @@ bool QueryReads::isGoodMapping() {
         bool ok = (minMapLength_ <= (unsigned int) (r->qe - r->qs) or minMapLengthPercent_ <= 100.0 * (r->qe - r->qs) / readPtr->seq.l);
         if (debug_) {
             filehandleDebugOut_ << "MAPPING\t" << readPtr->name.s
-                       << '\t' << readPtr->comment.s
+                       << '\t' << ((readPtr->comment.s != 0 and readPtr->comment.l > 0) ? readPtr->comment.s : "NA")
                        << '\t' << r->qs << "-" << r->qe
                        << '\t' << "+-"[r->rev]
                        << "\tpass:" << ok << '\n';

